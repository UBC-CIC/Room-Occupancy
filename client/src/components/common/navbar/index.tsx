import React from "react";
import { Menu, Segment, Grid, Button } from "semantic-ui-react";
import cicLogo from "../../../assets/images/cicLogo.png";
import { Link } from "react-router-dom";
import "../../../App.css";
import { useAuthenticator } from "@aws-amplify/ui-react";
import { handleSignOut } from "../../../pages/auth/Helpers";

export interface INavbarProps {}

interface NavBarPath {
  url: string;
  name: string;
}

export function Navbar(props: INavbarProps) {
  const { authStatus } = useAuthenticator((context) => [context.authStatus]);

  const navBarPaths: NavBarPath[] = [
    {
      url: "/",
      name: "Home",
    },
    {
      url: "/learn",
      name: "Learn",
    },
    {
      url: "/about",
      name: "About",
    },
    {
      url: "/analytics",
      name: "Analytics",
    },
  ];

  return (
    <nav className="navbarContainer">
      <Segment basic attached size="mini" fluid="true">
        <Grid>
          <Grid.Column width={6}>
            <img alt="logo" width={"300rem"} src={cicLogo} />
          </Grid.Column>
          <Grid.Column textAlign="center" width={5}>
            <Menu centered="true" basic="true" secondary>
              {navBarPaths.map((navPath) => {
                return (
                  <Link key={navPath.name} to={navPath.url}>
                    <Menu.Item role="menuitem">{navPath.name}</Menu.Item>
                  </Link>
                );
              })}
              {authStatus === "authenticated" && (
                <Link to="/admin_dashboard">
                  <Menu.Item role="menuitem">Admin Console</Menu.Item>
                </Link>
              )}
            </Menu>
          </Grid.Column>
          <Grid.Column textAlign="right" width={5}>
            {authStatus === "authenticated" ? (
              <Button color="blue" onClick={handleSignOut}>
                Logout
              </Button>
            ) : (
              <Link to="/login">
                <Button color="blue">Login</Button>
              </Link>
            )}
          </Grid.Column>
        </Grid>
      </Segment>
    </nav>
  );
}
