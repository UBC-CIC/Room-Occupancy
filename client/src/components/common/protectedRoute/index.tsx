import React from "react";
import { Navigate } from "react-router-dom";

export enum Roles {
  user = "user",
  admin = "admin",
}

interface Props {
  role: Roles;
  children: React.ReactNode;
}

export const ProtectedRoute = ({ role, children }: Props) => {
  if (role !== Roles.admin) {
    return <Navigate to="/" />;
  }

  return <>{children}</>;
};
