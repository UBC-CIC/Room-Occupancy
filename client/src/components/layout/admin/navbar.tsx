import React from "react";
import {
  Label,
  Segment,
  Grid,
  Icon,
  Header,
  Image,
  Dropdown,
  DropdownMenu,
} from "semantic-ui-react";
import "../../../App.css";
import { Link } from "react-router-dom";
import { useAppConfig } from "../../../providers/ConfigProvider";
import { handleSignOut } from "../../../pages/auth/Helpers";

interface Props {
  user: string;
  image: string | undefined;
}

export function AdminNavbar({ user, image }: Props) {
  const { config } = useAppConfig();
  return (
    <nav className="navbarContainer">
      <Segment basic attached size="mini" fluid="true">
        <Grid stackable={false} columns={3}>
          <Grid.Column width={4}>
            <img alt="logo" width={"300rem"} src={config.image} />
          </Grid.Column>
          <Grid.Column verticalAlign="middle" width={7}>
            <Header as="h3">Administator Console</Header>
          </Grid.Column>
          <Grid.Column textAlign="right" verticalAlign="middle" width={5}>
            <Link to="/admin_help">
              <Icon link size="large" name="question circle outline" />
            </Link>
            <Link to="/admin_settings">
              <Icon link size="large" name="setting" />
            </Link>
            <Icon link size="large" name="alarm" />
            {image === undefined ? (
              <Dropdown
                closeOnEscape
                text={user}
                className="icon"
                labeled
                button
                icon="user"
              >
                <DropdownMenu>
                  <Dropdown.Item
                    onClick={() => handleSignOut()}
                    text="Log Out"
                  />
                </DropdownMenu>
              </Dropdown>
            ) : (
              <Label size="large" as="a" basic image>
                <Image size="large" src={image} />
                {user}
              </Label>
            )}
          </Grid.Column>
        </Grid>
      </Segment>
    </nav>
  );
}
