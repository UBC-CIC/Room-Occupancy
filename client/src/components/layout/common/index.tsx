import React from "react";
import { Navbar } from "../../common/navbar";

type Props = {
  children: React.ReactNode;
};

export const CommonLayout = ({ children }: Props) => {
  return (
    <>
      <Navbar />
      {children}
    </>
  );
};
