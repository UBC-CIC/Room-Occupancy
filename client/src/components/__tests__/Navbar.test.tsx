import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Navbar } from "../common/navbar";
import { BrowserRouter as Router } from "react-router-dom";

describe("Navbar", () => {
  const renderNavbar = () => {
    return render(
      <Router>
        <Navbar />
      </Router>
    );
  };

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Navbar />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the navbar", () => {
    renderNavbar();
    expect(screen.getByRole("navigation")).toBeInTheDocument();
    expect(screen.getByAltText("logo")).toBeInTheDocument();
  });

  it("renders all navigation links", () => {
    const navPathNames = ["Home", "Learn", "About"];
    renderNavbar();
    const menuItems = screen.getAllByRole("menuitem");
    expect(menuItems).toHaveLength(4);

    navPathNames.map((name) => {
      expect(screen.getByText(name)).toBeInTheDocument();
    });
  });

  it("has correct navigation paths", () => {
    renderNavbar();
    expect(screen.getByText("Home").closest("a")).toHaveAttribute("href", "/");
    expect(screen.getByText("Learn").closest("a")).toHaveAttribute(
      "href",
      "/learn"
    );
    expect(screen.getByText("About").closest("a")).toHaveAttribute(
      "href",
      "/about"
    );
    expect(screen.getByText("Analytics").closest("a")).toHaveAttribute(
      "href",
      "/analytics"
    );
  });

  it("renders login and signup buttons", () => {
    renderNavbar();
    expect(screen.getByText("Login")).toBeInTheDocument();
    expect(screen.getByText("SignUp")).toBeInTheDocument();
  });
});
