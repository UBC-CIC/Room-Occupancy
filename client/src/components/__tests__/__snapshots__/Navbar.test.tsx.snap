// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbarContainer"
  >
    <div
      class="ui mini basic attached segment"
      fluid="true"
    >
      <div
        class="ui grid"
      >
        <div
          class="six wide column"
        >
          <img
            alt="logo"
            src="test-file-stub"
            width="300rem"
          />
        </div>
        <div
          class="center aligned five wide column"
        >
          <div
            basic="true"
            centered="true"
            class="ui secondary menu"
          >
            <a
              href="/"
            >
              <div
                class="item"
                role="menuitem"
              >
                Home
              </div>
            </a>
            <a
              href="/learn"
            >
              <div
                class="item"
                role="menuitem"
              >
                Learn
              </div>
            </a>
            <a
              href="/about"
            >
              <div
                class="item"
                role="menuitem"
              >
                About
              </div>
            </a>
            <a
              href="/analytics"
            >
              <div
                class="item"
                role="menuitem"
              >
                Analytics
              </div>
            </a>
          </div>
        </div>
        <div
          class="right aligned five wide column"
        >
          <div
            class="ui buttons"
          >
            <a
              href="/login"
            >
              <button
                class="ui button"
              >
                Login
              </button>
            </a>
            <div
              class="or"
              data-text="or"
            />
            <a
              href="/signup"
            >
              <button
                class="ui blue button"
              >
                SignUp
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
